packages/blade/src/components/Input/BaseInput/BaseInput.tsx

```diff
@@ -1,73 +1,166 @@
-   * Type of Input Field to be rendered.
+   * Error text for the input
    *
-   * @default text
-   */
-  type?: 'text' | 'telephone' | 'email' | 'url' | 'numeric' | 'search';
-  /**
-   * Used to set the default value of input field when it's uncontrolled
+   * Renders when `validationState` is set to 'error'
    */
-  defaultValue?: string;
+  errorText?: string;
   /**
-   * The name of the input field.
+   * success text for the input
    *
-   * Useful in form submissions
+   * Renders when `validationState` is set to 'success'
    */
-  name?: string;
+  successText?: string;
   /**
-   * The callback function to be invoked when the value of the input field changes
+   * If `error`, the input is marked as invalid,
+   * and `invalid` attribute will be added
+   *
+   * If `success`, the input is marked as invalid,
+   *
    */
-  onChange?: OnChange;
+  validationState?: 'success' | 'error' | 'none';
+};
+
+export type BaseInputProps = InputLabelProps &
+  InputValidationProps & {
+    /**
+     * ID that will be used for accessibility
+     */
+    id: string;
+    /**
+     * Placeholder text to be displayed inside the input field
+     */
+    placeholder?: string;
+    /**
+     * Type of Input Field to be rendered.
+     *
+     * @default text
+     */
+    type?: 'text' | 'telephone' | 'email' | 'url' | 'numeric' | 'search';
+    /**
+     * Used to set the default value of input field when it's uncontrolled
+     */
+    defaultValue?: string;
+    /**
+     * The name of the input field.
+     *
+     * Useful in form submissions
+     */
+    name?: string;
+    /**
+     * The callback function to be invoked when the value of the input field changes
+     */
+    onChange?: OnChange;
+    /**
+     * Used to turn the input field to controlled so user can control the value
+     */
+    value?: string;
+    /**
+     * Used to disable the input field
+     */
+    isDisabled?: boolean;
+    /**
+     * If true, the input is marked as required, and `required` attribute will be added
+     */
+    isRequired?: boolean;
+  };
+
+const useInput = ({
+  value,
+  defaultValue,
+  onChange,
+}: Pick<BaseInputProps, 'value' | 'defaultValue' | 'onChange'>): {
+  handleOnChange: HandleOnChange;
+} => {
+  if (value && defaultValue) {
+    throw new Error(
+      `[Blade Input]: Either 'value' or 'defaultValue' shall be passed. This decides if the input field is controlled or uncontrolled`,
+    );
+  }
+
+  const handleOnChange: HandleOnChange = React.useCallback(
+    ({ inputName, inputValue }) => {
+      let _inputValue = '';
+
+      if (getPlatformType() === 'react-native' && typeof inputValue === 'string') {
+        _inputValue = inputValue;
+      } else if (typeof inputValue !== 'string') {
+        // it's weird but TS forced me to write this much code where I could have just done "getPlatformType() === 'react-native' ? inputValue : inputValue?.target.value" :(
+        _inputValue = inputValue?.target.value ?? '';
+      }
```

Comment:
Does this look a bit more cleaner? 

> also uplift the `const isReactNative = getPlatformType() === 'react-native';` check to component body, no need to check on each and every keystroke. 

```ts
      const isValueString = typeof inputValue === 'string';
      if (isReactNative && isValueString) {
        _inputValue = inputValue;
      }
      if (!isValueString) {
        // it's weird but TS forced me to write this much code where I could have just done "getPlatformType() === 'react-native' ? inputValue : inputValue?.target.value" :(
        _inputValue = inputValue?.target.value ?? '';
      }
```

Reply:
don't mind doing that but honestly i don't see any value of doing that. it's just moving things around 1 level up where we don't even get any other advantage. if `getPlatformType` and `typeof` operation would have been very heavy i would have done that but they are very very very light weight operations so wouldn't make any diff by moving up at the cost of readability.